### PR TITLE
MersenneTwister: implement generation of jump-ahead polynomials

### PR DESCRIPTION
**Pull request by rfourquet**
This completes PR #12498, which provided the jump-ahead functionality, by allowing to specify an arbitrary number of steps instead of the hard-coded 10^20. This code may very well be better-suited for a package, but it's so tiny (roughly the same size as the original code from #12498) that I thought I could push it here as well.

This is WIP as I'm considering evolving the `randjump` API, e.g. deprecating the method `randjump(rng::MersenneTwister, jumps, jumppoly::AbstractString)`. Also I think I would like to have `randjump(rng, steps)` (and `randjump!(rng, steps)`) but this is incompatible with `randjump(rng, jumps)`.

### CODE REVIEW

File: base/dSFMT.jl [hunk 1]
```diff
@@ -79,20 +81,94 @@ function dsfmt_fill_array_close_open!(s::DSFMT_state, A::Ptr{Float64}, n::Int)
...
+    e < 0 && throw(DomainError())
+    foldl(GF2X(1), Base.ndigits0z(e, 2)-1:-1:0) do f, i
+        tstbit(e, i) ?
+            mulxmod!(sqrmod!(f, m), m) :
+            sqrmod!(f, m)
+    end
+end
+
+"Cached jump polynomials for `MersenneTwister`."
+# hack: contains Poly19937 at key -1 (as it can not be initialized at compile time)
```
Comment (tkelman) [inline base/dSFMT.jl]: need to double check this, but having the comment between the docstring and the const might mean the docstring is not associated with the const?

File: base/random.jl [hunk 2]
```diff
@@ -112,7 +112,34 @@ function srand(r::MersenneTwister, seed::Vector{UInt32})
...
-function randjump(mt::MersenneTwister, jumps::Integer, jumppoly::AbstractString)
+
+"""
+    randjump(r::MersenneTwister, jumps, steps=10^20) -> Vector{MersenneTwister}
+
+Create an array of size `jumps` of initialized `MersenneTwister` RNG
+objects corresponding to different substreams of `r`, using the
+"jump-ahead" method. The first element of this array is the RNG `r`
+given as a parameter; each other RNG element is initialized to a state
+equivalent to what would be that of the previous element if it was
```
Comment (tkelman) [inline base/random.jl]: this is kind of confusing phrasing

File: base/random.jl [hunk 3]
```diff
@@ -112,7 +112,34 @@ function srand(r::MersenneTwister, seed::Vector{UInt32})
...
+moved forward by `steps` steps (one such "step" corresponds to the
+generation of two `Float64` numbers, but no numbers are actually
+generated). This allows to obtain multiple streams of random numbers
+from one intitial state, while still benefiting from strong
```
Comment (tkelman) [inline base/random.jl]: "allows obtaining," and initial misspelled

File: base/random.jl [hunk 4]
```diff
@@ -112,7 +112,34 @@ function srand(r::MersenneTwister, seed::Vector{UInt32})
...
+statistical properties (provided of course that the substreams don't
+overlap).
+
+    randjump(r::MersenneTwister, jumps, jumppoly::AbstractString) -> Vector{MersenneTwister}
```
Comment (tkelman) [inline base/random.jl]: this should be a separate docstring

File: test/random.jl [hunk 5]
```diff
@@ -390,23 +390,25 @@ let mta = MersenneTwister(42), mtb = MersenneTwister(42)
...
     mts = randjump(mta, size, jump25000)
     @test length(mts) == 4
-
-    tmp = zeros(Float64, size*step)
-    for j in 1:step
-        for k in 1:size
-            tmp[j + (k-1) * step] = rand(mts[k], Float64)
-        end
+    for x in (rand(mts[k], Float64) for j=1:step, k=1:size)
+       @test rand(mtb, Float64) == x
```
Comment (tkelman) [inline test/random.jl]: indent slightly off

**Comment by wildart** [comment]
You can make `jumps` parameter in `randjump` to default to 1. And add `steps` as keyword argument so
```julia
randjump(rng, 10) # ten jumps with default steps
randjump(rng, steps=10) # one jump with 10 steps
randjump(rng, 10, steps=10) # ten jumps with 10 steps
```

It's the same thing for `randjump!`.

File: base/dSFMT.jl (hunk 1 above)
Comment (rfourquet) [inline base/dSFMT.jl]: I just checked and it works! but I'll move anyway the comment after the definition.

File: base/random.jl (hunk 2 above)
Comment (rfourquet) [inline base/random.jl]: Requesting help for improvements here...

Comment (tkelman) [inline base/random.jl]: can you just say "the previous element's state moved forward by `steps` steps" ? what is "equivalent to what would be that of" trying to convey here?

Comment (rfourquet) [inline base/random.jl]: Ah yes much better! (I guess I wanted to make clear that there is no sharing of state (storage) between the two elements).

File: base/random.jl [hunk 6]
```diff
@@ -112,7 +112,35 @@ function srand(r::MersenneTwister, seed::Vector{UInt32})
...
+overlap).
+"""
+randjump(r::MersenneTwister, jumps::Integer, steps::Integer=big(10)^20) =
+    randjump(r, jumps, dSFMT.calc_jump(steps))
+
+"""
+    randjump(r::MersenneTwister, jumps, jumppoly::AbstractString) -> Vector{MersenneTwister}
+
+Similar to `randjump(r, jumps, steps)` where the number of steps is determined
+by a jump polynomial `jumppoly` in ``GF(2)[X]`` encoded as an hexadecimal string.
```
Comment (rfourquet) [inline base/random.jl]: Updated.

**Comment by rfourquet** [comment]
I think I agree with your proposed API for `randjum`. I will remove the WIP, as I now think it's better to discuss an updated API in a new PR, which will depend on wheter this present PR gets merged or not.

**Comment by rfourquet** [comment]
This is freshly rebased. Barring objections, I will just merge this: this adds only a little bit of code, and adds the missing half of the `randjump` functionality. And with `Random` moving to "stdlib", it makes even more sense, as opposed to putthing this into a separate package. Moreover, this will be useful as a complement of #25129.

**Pull request merged**